pkg/remote/write.go

```diff
@@ -0,0 +1,256 @@
+	tick := time.NewTicker(w.checkInterval)
+	defer tick.Stop()
+
+	// This is true when a retriable error happened. As the messages we read
+	// from Pulsar can be for any tenant, we need to block all consumption, to
+	// avoid overloading the adapter.
+	//
+	// TODO: Figure out a better way though e.g. per tenant queues.
+	errRemoteWriteRetryable := false
+	blockingSampleCh := make(chan pulsar.ReceivedSample)
+
+	sample := func() chan pulsar.ReceivedSample {
+		// don't return new samples if err has happend
+		if errRemoteWriteRetryable {
+			return blockingSampleCh
+		}
+		return sampleCh
+	}
+
+	var samplesSinceCheck int
+	var deadlineCheck time.Time
+
+receive:
+	for {
+		// wait for either tick or incoming sample, handle closed context
+		select {
+		case <-ctx.Done():
+			break receive
+		case <-tick.C:
+			// do nothing
+		case sample := <-sample():
+			tenantID := mcontext.TenantIDFromContext(sample.Context)
+			if _, ok := w.samplePerTenantID[tenantID]; !ok {
+				w.samplePerTenantID[tenantID] = []pulsar.ReceivedSample{}
+				w.deadlinePerTenantID[tenantID] = time.Now().Add(w.BatchMaxDelay)
+			}
+
+			w.samplePerTenantID[tenantID] = append(w.samplePerTenantID[tenantID],
+				sample,
+			)
+			samplesSinceCheck += 1
+		}
+
+		// continue reading samples from channel, if no errors has happened, we
+		// didn't hit the check deadline and the samples received are smaller
+		// than the BatchSize:
+		if !errRemoteWriteRetryable &&
+			time.Now().Before(deadlineCheck) &&
+			samplesSinceCheck < w.BatchSize {
+			continue
+		}
+
+		// reset check conditions
+		errRemoteWriteRetryable = false
+		samplesSinceCheck = 0
+		deadlineCheck = time.Now().Add(w.checkInterval)
+
+		// loop through tenants and find metrics to send
+		for tenantID, samples := range w.samplePerTenantID {
+			logger := tenantIDLogger(w.logger, tenantID)
+
+			batchSizeReached := len(samples) >= w.BatchSize
+			pastDeadline := w.deadlinePerTenantID[tenantID].Before(time.Now())
+
+			if !batchSizeReached && !pastDeadline {
+				continue
+			}
+
+			req := samplesToProto(samples)
+
+			data, err := proto.Marshal(req)
+			if err != nil {
+				return err
+			}
+
+			compressed := snappy.Encode(nil, data)
+
+			if err := client.Store(mcontext.ContextWithTenantID(ctx, tenantID), compressed); err != nil {
+				errRec := &remote.RecoverableError{}
+				if errors.As(err, errRec) {
+					_ = level.Warn(w.logger).Log("msg", "failed remote_write request, will retry", "error", err)
+					errRemoteWriteRetryable = true
+					continue
+				}
+
+				// non-recoverable errors still require all messages to be
+				// acked, as we otherwise would get them redelivered through
+				// pulsar
+				_ = level.Error(w.logger).Log("msg", "failed remote_write reqeust", "error", err)
+
+			} else {
+				_ = level.Debug(logger).Log(
+					"msg", "remote_write request succesful",
+					"sample_count", len(samples),
+				)
+			}
+
+			// Ack the messages
```

Comment:
Maybe expand this comment to indicate we're acking successes and non-recoverable errors (similar to the above comment)?

Reply:
Good point d2803dd